### PR TITLE
User management: C.R.U.D, search

### PR DESCRIPTION
**Pull request by trangnvm**
- Files change :
app/User.php
config/app.php
routes/web.php
app/Http/Controllers/User/UserManagementController.php
app/Rules/CheckEmailExists.php
app/Rules/CheckEmailUpdate.php
database/migrations/2018_04_09_015618_add_role_to_user_table.php
resources/views/users/create.blade.php
resources/views/users/edit.blade.php
resources/views/users/index.blade.php

### CODE REVIEW

File: resources/views/users/index.blade.php [hunk 1]
```diff
@@ -0,0 +1,119 @@
+<!DOCTYPE html>
+<html>
+    <head>
+        <title>User Management</title>
+        <link rel="stylesheet" href="//netdna.bootstrapcdn.com/bootstrap/3.0.0/css/bootstrap.min.css">
```
Comment (ducla5) [inline resources/views/users/index.blade.php]: Laravel use bootstrab 4.0 by sass. Don't import it manually.
use  `<link href="{{ asset('css/app.css') }}" rel="stylesheet">`
Separate style to a CSS file

File: resources/views/users/create.blade.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
+<!DOCTYPE html>
+<html>
+    <head>
+        <title>User Management</title>
+        <link rel="stylesheet" href="//netdna.bootstrapcdn.com/bootstrap/3.0.0/css/bootstrap.min.css">
```
Comment (ducla5) [inline resources/views/users/create.blade.php]: Laravel use bootstrab 4.0 by sass. Don't import it manually.
use  <link href="{{ asset('css/app.css') }}" rel="stylesheet">

File: resources/views/users/edit.blade.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
+<!DOCTYPE html>
+<html>
+    <head>
+        <title>User Management</title>
+        <link rel="stylesheet" href="//netdna.bootstrapcdn.com/bootstrap/3.0.0/css/bootstrap.min.css">
```
Comment (ducla5) [inline resources/views/users/edit.blade.php]: Laravel use bootstrab 4.0 by sass. Don't import it manually.
use  <link href="{{ asset('css/app.css') }}" rel="stylesheet">

File: resources/views/users/index.blade.php [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+                                    @elseif($value->role == 1)
+                                        Admin
+                                    @elseif($value->role == 2)
+                                        User
+                                    @endif
+                                </td>
+
+                                <td class="vertical-text">
+                                    @if ($value->role != 0)
+                                        <a class="btn btn-small btn-info" href="{{ URL::to('users/' . $value->id . '/edit') }}">Edit</a>
```
Comment (ducla5) [inline resources/views/users/index.blade.php]: use route name
-> `<?php echo route('users.edit', ['id' => $value->id]); ?>`

File: resources/views/users/index.blade.php [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+                                        {{ Form::checkbox('delete_id[]', $value->id, null, ['class' => 'pull-right']) }}
+                                    @endif                                    
+                                </td>
+                            </tr>
+                        @endforeach
+                    </tbody>
+                </table>
+
+                <div class="div-button-footer">
+                    <a class="btn btn-small btn-success" href="{{ URL::to('users/create') }}">Create New User</a>
```
Comment (ducla5) [inline resources/views/users/index.blade.php]: use `{{ route('users.create') }}`

File: app/Http/Controllers/User/UserManagementController.php [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+
+        // redirect
+        Session::flash('message', 'Successfully deleted the user!');
+        return Redirect::to('users');
+    }
+
+    /**
+     * Remove the specified resources from storage.
+     *
+     * @return \Illuminate\Http\Response
```
Comment (ducla5) [inline app/Http/Controllers/User/UserManagementController.php]: wrong return type in comment
@return \Illuminate\Http\RedirectResponse

File: app/Http/Controllers/User/UserManagementController.php [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+                return $e->getMessage();
+            }
+        }
+    }
+
+    /**
+     * Remove the specified resource from storage.
+     *
+     * @param  int  $id
+     * @return \Illuminate\Http\Response
```
Comment (ducla5) [inline app/Http/Controllers/User/UserManagementController.php]: wrong return type in comment

File: app/Http/Controllers/User/UserManagementController.php [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+        }
+        return view('users.edit', ['user' => $user]);
+    }
+
+    /**
+     * Update the specified resource in storage.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @param  int  $id
+     * @return \Illuminate\Http\Response
```
Comment (ducla5) [inline app/Http/Controllers/User/UserManagementController.php]: wrong return type in comment

File: app/Http/Controllers/User/UserManagementController.php [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+     */
+    public function update(Request $request, $id)
```
Comment (ducla5) [inline app/Http/Controllers/User/UserManagementController.php]: unused parametter $request

File: resources/views/users/index.blade.php (hunk 1 above)
Comment (trangnvm) [reply resources/views/users/index.blade.php]: Done

File: resources/views/users/create.blade.php (hunk 2 above)
Comment (trangnvm) [reply resources/views/users/create.blade.php]: Done

File: resources/views/users/edit.blade.php (hunk 3 above)
Comment (trangnvm) [reply resources/views/users/edit.blade.php]: Done

File: resources/views/users/index.blade.php (hunk 5 above)
Comment (trangnvm) [reply resources/views/users/index.blade.php]: Done

File: resources/views/users/index.blade.php (hunk 4 above)
Comment (trangnvm) [reply resources/views/users/index.blade.php]: Done

File: app/Http/Controllers/User/UserManagementController.php [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+    /**
+     * Remove the specified resources from storage.
+     *
+     * @return \Illuminate\Http\Response
+     */
+    public function list_delete()
+    {
+        // delete
+        $ids_to_delete = Input::get('delete_id');
+        $user = User::delete_users($ids_to_delete);
```
[the remark this answers is not in the document]
Comment (trangnvm) [reply app/Http/Controllers/User/UserManagementController.php]: Done

File: app/Http/Controllers/User/UserManagementController.php [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+        return view('users.create');
+    }
+
+    /**
+     * Store a newly created resource in storage.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @return \Illuminate\Http\Response
+     */
+    public function store(Request $request)
```
[the remark this answers is not in the document]
Comment (trangnvm) [reply app/Http/Controllers/User/UserManagementController.php]: Done

File: app/Http/Controllers/User/UserManagementController.php [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+{
+    /**
+     * Display a listing of the resource.
+     *
+     * @return \Illuminate\Http\Response
+     */
+    public function index()
+    {
+        $users = User::paginate(20);
+
```
[the remark this answers is not in the document]
Comment (trangnvm) [reply app/Http/Controllers/User/UserManagementController.php]: Done

File: app/Http/Controllers/User/UserManagementController.php [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+        }
+    }
+
+    /**
+     * Display the specified resource.
+     *
+     * @param  int  $id
+     * @return \Illuminate\Http\Response
+     */
+    public function show($id)
```
[the remark this answers is not in the document]
Comment (trangnvm) [reply app/Http/Controllers/User/UserManagementController.php]: Done

File: app/Http/Controllers/User/UserManagementController.php [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+     * @return \Illuminate\Http\Response
+     */
+    public function destroy($id)
+    {
+        // delete
+        $user = User::find($id);
+        $user -> delete();
+
+        // redirect
+        Session::flash('message', 'Successfully deleted the user!');
```
[the remark this answers is not in the document]
Comment (trangnvm) [reply app/Http/Controllers/User/UserManagementController.php]: Done

File: app/Rules/CheckEmailExists.php [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+use DB;
+
+class CheckEmailExists implements Rule
+{
+    /**
+     * Create a new rule instance.
+     *
+     * @return void
+     */
+    public function __construct()
```
[the remark this answers is not in the document]
Comment (trangnvm) [reply app/Rules/CheckEmailExists.php]: Done

File: database/migrations/2018_04_09_015618_add_role_to_user_table.php [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    /**
+     * Reverse the migrations.
+     *
+     * @return void
+     */
+    public function down()
+    {
+        //
```
[the remark this answers is not in the document]
Comment (trangnvm) [reply database/migrations/2018_04_09_015618_add_role_to_user_table.php]: Done

File: routes/web.php [hunk 17]
```diff
@@ -14,3 +14,7 @@
 Route::get('/', function () {
     return view('welcome');
 });
+
+Route::resource('users', 'User\UserManagementController');
+Route::post('deleteUsers', 'User\UserManagementController@list_delete')->name('deleteUsers');
+Route::get('search', 'User\UserManagementController@search_users')->name('search');
```
[the remark this answers is not in the document]
Comment (trangnvm) [reply routes/web.php]: Done

File: app/Http/Controllers/User/UserManagementController.php (hunk 6 above)
Comment (trangnvm) [reply app/Http/Controllers/User/UserManagementController.php]: Done

File: app/Http/Controllers/User/UserManagementController.php (hunk 8 above)
Comment (trangnvm) [reply app/Http/Controllers/User/UserManagementController.php]: Done

File: app/Http/Controllers/User/UserManagementController.php (hunk 9 above)
Comment (trangnvm) [reply app/Http/Controllers/User/UserManagementController.php]: Done

**Pull request merged**